session/session.go

```diff
@@ -1528,14 +1528,16 @@ func (s *session) ParseWithParams(ctx context.Context, sql string, args ...inter
 	return stmts[0], nil
 }
 
-// ParseWithParamsInternal is same as ParseWithParams except set `s.sessionVars.InRestrictedSQL = true`
+// ParseWithParamsInternal is same as ParseWithParams except use an internal session
 func (s *session) ParseWithParamsInternal(ctx context.Context, sql string, args ...interface{}) (ast.StmtNode, error) {
-	origin := s.sessionVars.InRestrictedSQL
-	s.sessionVars.InRestrictedSQL = true
-	defer func() {
-		s.sessionVars.InRestrictedSQL = origin
-	}()
-	return s.ParseWithParams(ctx, sql, args...)
+	tmp, err := s.sysSessionPool().Get()
+	if err != nil {
+		return nil, err
+	}
+	defer s.sysSessionPool().Put(tmp)
+	se := tmp.(*session)
```

Comment:
The design of this function is problematic, the parse and the execute is not using the same session.
If the SQL has some side-effect, for example, a warning message, it's left on the pool's session.

A better solution is a callback style API:

```
func (s *session) runWithInternalSQLExec(fn func(exec sqlexec.RestrictedSQLExec) error) {
    tmp, err := s.sysSessionPool().Get()
	if err != nil {
		return nil, err
	}
	defer s.sysSessionPool().Put(tmp)
    se := tmp.(*session)
    return fn(se)
}
```

The call use it like this:

```
se.runWithInternalSQLExec(func(exec *exec.RestrictedSQLExec) {
    stmt := exec.ParseWithParamsInternal(...)
    exec.Exec(stmt)
    ...
})
```

And you can finally wrap a simpler API like:

```
func (s *session) ExecRestrictedSQL(sql string) {
    s.runWithInternalSQLExec(func(exec sqlexec.RestrictedSQLExec) {
        stmt := exec.ParseWithParamsInternal(...)
        exec.Exec(stmt)
        ...        
    })
}
```

The simplied API is more convenient, but its design is not secure.

Reply:
> But we didn't use same session before. And I don't know it's a bug or feature 😭.

Before a previous refactoring, they're in the same session.
The data race is introduced after that refactoring

Reply:
Fix the data race fisrt, it may affects 5.4.0.